libass/ass_bitmap.c

```diff
@@ -679,32 +698,64 @@ void be_blur_c(uint8_t *buf, intptr_t w,
             temp1 = old_sum + temp2;
             old_sum = temp2;
 
-            temp2 = col_pix_buf[x] + temp1;
-            col_pix_buf[x] = temp1;
-            col_sum_buf[x] = temp2;
+            temp2 = col_pix_buf[x-1] + temp1;
+            col_pix_buf[x-1] = temp1;
+            dst[x-1] = (col_sum_buf[x-1] + temp2) >> 4;
+            col_sum_buf[x-1] = temp2;
         }
+        temp1 = old_sum + old_pix;
+        temp2 = col_pix_buf[x-1] + temp1;
+        col_pix_buf[x-1] = temp1;
+        dst[x-1] = (col_sum_buf[x-1] + temp2) >> 4;
+        col_sum_buf[x-1] = temp2;
     }
 
-    for (y = 2; y < h; y++) {
-        src=buf+y*stride;
+    {
         dst=buf+(y-1)*stride;
+        for (x = 0; x < w; x++)
+            dst[x] = (col_sum_buf[x] + col_pix_buf[x]) >> 4;
+    }
+}
 
-        x = 2;
-        old_pix = src[x-1];
-        old_sum = old_pix + src[x-2];
-        for ( ; x < w; x++) {
-            temp1 = src[x];
-            temp2 = old_pix + temp1;
-            old_pix = temp1;
-            temp1 = old_sum + temp2;
-            old_sum = temp2;
+void be_blur_pre(uint8_t *buf, intptr_t w, intptr_t h, intptr_t stride)
+{
+    for (int y = 0; y < h; ++y)
+    for (int x = 0; x < w; ++x)
+        buf[y * stride + x] = (buf[y * stride + x] * 64 + 127) / 255;
+}
 
-            temp2 = col_pix_buf[x] + temp1;
-            col_pix_buf[x] = temp1;
-            dst[x-1] = (col_sum_buf[x] + temp2) >> 4;
-            col_sum_buf[x] = temp2;
-        }
-    }
+void be_blur_post(uint8_t *buf, intptr_t w, intptr_t h, intptr_t stride)
+{
+    for (int y = 0; y < h; ++y)
+    for (int x = 0; x < w; ++x)
+        buf[y * stride + x] = (buf[y * stride + x] * 255 + 32) / 64;
+}
+
+/*
+ * To find these values, simulate blur on the border between two
+ * half-planes, one zero-filled (background) and the other filled
+ * with the maximum supported value (foreground). Keep incrementing
+ * the \be argument. The necessary padding is the distance by which
+ * the blurred foreground image extends beyond the original border
+ * and into the background. Initially it increases along with \be,
+ * but very soon it grinds to a halt. At some point, the blurred
+ * image actually reaches a stationary point and stays unchanged
+ * forever after, simply _shifting_ by one pixel for each \be
+ * step--moving in the direction of the non-zero half-plane
+ * and thus decreasing the necessary padding (although the large padding
+ * is still needed for intermediate results). Hence the necessary
+ * padding has a global maximum, which this simulation determines.
+ * In practice, images are finite rather than infinite like half-planes,
+ * but this can only decrease the required padding. Half-planes filled
+ * with extreme values are the theoretical limit of the worst case.
+ */
+int be_padding(int be)
+{
+    static const int limits[] = {0, 1, 2, 6, INT_MAX};
+    int i = 0;
+    while (be > limits[i])
+        i++;
+    return i;
```

Comment:
If this code is deemed too complicated, we could use a constant radius that’s enough for any \be value ever. But doing so would increase our memory usage and slow down various operations dealing with bitmaps. (Potentially noticeable when there are large amounts of very small bitmaps.)